src/async_impl/decoder.rs

```diff
@@ -0,0 +1,297 @@
+    PlainText(ReadableChunks<TPollChunk>),
+    /// A `Gzip` decoder will uncompress the gziped response content before returning it.
+    Gzip(gzip::Decoder<Peeked<ReadableChunks<TPollChunk>>>),
+    /// An error occured reading the Gzip header, so return that error
+    /// when the user tries to read on the `Response`.
+    Errored(Option<io::Error>),
+}
+
+enum DecoderInnerFuture<TPollChunk> {
+    Empty,
+    PendingGzip(ReadableChunks<TPollChunk>)
+}
+
+struct Peeked<TReadableChunks> {
+    peeked: Option<u8>,
+    inner: TReadableChunks,
+}
+
+impl<TPollChunk> Decoder<TPollChunk> 
+    where TPollChunk: PollChunk
+{
+    #[inline]
+    fn errored(err: Option<io::Error>) -> Decoder<TPollChunk> {
+        Decoder {
+            inner_future: DecoderInnerFuture::Empty,
+            inner: Some(DecoderInner::Errored(err))
+        }
+    }
+
+    #[inline]
+    fn plain_text(body: ReadableChunks<TPollChunk>) -> Decoder<TPollChunk> {
+        Decoder {
+            inner_future: DecoderInnerFuture::Empty,
+            inner: Some(DecoderInner::PlainText(body))
+        }
+    }
+
+    #[inline]
+    fn gzip(mut body: ReadableChunks<TPollChunk>) -> Decoder<TPollChunk> {
+        Decoder {
+            inner_future: DecoderInnerFuture::PendingGzip(body),
+            inner: None
+        }
+    }
+}
+
+impl<TPollChunk> Read for Decoder<TPollChunk> 
+    where TPollChunk: PollChunk
+{
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if let Some(ref mut inner) = self.inner {
+            match *inner {
+                DecoderInner::PlainText(ref mut body) => body.read(buf),
+                DecoderInner::Gzip(ref mut decoder) => decoder.read(buf),
+                DecoderInner::Errored(ref mut err) => {
+                    Err(err.take().unwrap_or_else(previously_errored))
+                }
+            }
+        }
+        else {
+            match self.inner_future.poll() {
+                Ok(Async::Ready(inner)) => {
+                    self.inner = Some(inner);
+                    self.read(buf)
+                },
+                Ok(Async::NotReady) => Err(io::ErrorKind::WouldBlock.into()),
+                Err(e) => Err(e)
+            }
+        }
+    }
+}
+
+impl<TPollChunk> Future for DecoderInnerFuture<TPollChunk>
+    where TPollChunk: PollChunk
+{
+    type Item = DecoderInner<TPollChunk>;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // libflate does a read_exact([0; 2]), so its impossible to tell
+        // if the stream was empty, or truly had an UnexpectedEof.
+        // Therefore, we need to peek a byte to make check for EOF first.
+        let mut peek = [0];
+        let read = match *self {
+            DecoderInnerFuture::PendingGzip(ref mut body) => {
+                match body.read(&mut peek) {
+                    Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => return Ok(Async::NotReady),
+                    read @ _ => read
```

Comment:
This condition doesn't look cool to me. Can you explain? Is it same as `x => x`?